modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/namehandler/CoverageNameHandlerSPI.java

```diff
@@ -0,0 +1,69 @@
+/*
+ * GeoTools - The Open Source Java GIS Toolkit
+ * http://geotools.org
+ *
+ * (C) 2016, Open Source Geospatial Foundation (OSGeo)
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation;
+ * version 2.1 of the License.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ */
+package org.geotools.gce.imagemosaic.namehandler;
+
+import java.util.Map;
+
+import javax.imageio.ImageReader;
+
+import org.geotools.coverage.grid.io.GridCoverage2DReader;
+
+/**
+ * CoverageNameHandler SPI interface. Allows to instantiate {@link CoverageNameHandler}s
+ * based on optional input source object and customization properties.
+ * 
+ * Implementations are pluggable. 
+ * In addition to implementing this interface, this service file should be defined:
+ * </p>
+ *
+ * <p>
+ * <code>META-INF/services/org.geotools.gce.imagemosaic.namehandler.CoverageNameHandlerSPI</code>
+ * </p>
+ *
+ * <p>
+ * The file should contain a single line which gives the full name of the implementing class.
+ * </p>
+ *
+ * <p>
+ * example:<br/>
+ * <code>e.g.
+ * org.geotools.gce.imagemosaic.namehandler.DefaultCoverageNameHandlerSPI</code>
+ * </p>
+ *
+ * 
+ */
+public interface CoverageNameHandlerSPI {
+
+    /** Create a CoverageNameHandler instance */
+    CoverageNameHandler create(Object object, Map<String,String> properties);
+
+    /** The {@link CoverageNameHandler} instance */
+    public interface CoverageNameHandler {
```

Comment:
Please make this one public, the hardest to read parts of the mosaic code are so because of overuse of inner classes and inner interfaces